File: src/agentprobe/judge.py
from __future__ import annotations

import json

import openai

from pydantic import ConfigDict, Field

from .data.common import AgentProbeModel
from .data.rubrics import JudgeConfig, Rubric


class JudgeDimensionScore(AgentProbeModel):
    model_config = ConfigDict(extra="forbid")

    reasoning: str
    evidence: list[str] = Field(default_factory=list)
    score: float


class RubricScore(AgentProbeModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    dimensions: dict[str, JudgeDimensionScore] = Field(default_factory=dict)
    overall_notes: str = ""
    passed: bool = Field(alias="pass")

    def validate_dimensions(self, rubric: Rubric) -> None:
        expected_ids = {dimension.id for dimension in rubric.dimensions}
        actual_ids = set(self.dimensions)

        missing_ids = sorted(expected_ids - actual_ids)
        extra_ids = sorted(actual_ids - expected_ids)

        if missing_ids or extra_ids:
            details: list[str] = []
            if missing_ids:
                details.append(f"missing dimensions: {', '.join(missing_ids)}")
            if extra_ids:
                details.append(f"unexpected dimensions: {', '.join(extra_ids)}")
            raise ValueError(
                "Judge output did not match rubric dimensions: " + "; ".join(details)
            )


def _judge_json_schema(rubric: Rubric) -> dict[str, object]:
    dimension_score_schema: dict[str, object] = {
        "type": "object",
        "properties": {
            "reasoning": {
                "type": "string",
                "description": "Concise reasoning for the assigned score.",
            },
            "evidence": {
                "type": "array",
                "items": {"type": "string"},
                "description": "Short evidence snippets or observations from the transcript.",
            },
            "score": {
                "type": "number",
                "description": "Numeric score for the rubric dimension.",
            },
        },
        "required": ["reasoning", "evidence", "score"],
        "additionalProperties": False,
    }

    dimensions_schema: dict[str, object] = {
        "type": "object",
        "properties": {
            dimension.id: {
                **dimension_score_schema,
                "description": f"Score details for rubric dimension `{dimension.id}`.",
            }
            for dimension in rubric.dimensions
        },
        "required": [dimension.id for dimension in rubric.dimensions],
        "additionalProperties": False,
    }

    return {
        "type": "object",
        "properties": {
            "dimensions": dimensions_schema,
            "overall_notes": {
                "type": "string",
                "description": "Short overall summary of strengths and failures.",
            },
            "pass": {
                "type": "boolean",
                "description": "Whether the evaluated response passes the rubric.",
            },
        },
        "required": ["dimensions", "overall_notes", "pass"],
        "additionalProperties": False,
    }


def _judge_instructions(rubric: Rubric, schema: dict[str, object]) -> str:
    dimension_ids = ", ".join(sorted(dimension.id for dimension in rubric.dimensions))
    return (
        "You are an expert rubric judge. Evaluate only the provided response.\n\n"
        f"{rubric.to_prompt_markdown()}\n\n"
        "Return structured output matching the requested schema exactly. "
        f"The `dimensions` object must contain exactly these rubric dimension ids: {dimension_ids}.\n\n"
        "JSON schema:\n"
        f"{json.dumps(schema, indent=2, sort_keys=True)}"
    )


def _judge_config(rubric: Rubric) -> JudgeConfig:
    if rubric.judge is None:
        raise ValueError("Rubric is missing judge configuration.")
    if rubric.judge.provider != "openai":
        raise ValueError(
            f"judge.py only supports OpenAI judges, got: {rubric.judge.provider}"
        )
    return rubric.judge


def _extract_response_text(response: object) -> str:
    direct_text = getattr(response, "output_text", None)
    if isinstance(direct_text, str) and direct_text.strip():
        return direct_text.strip()

    output = getattr(response, "output", None)
    if isinstance(output, list):
        chunks: list[str] = []
        for item in output:
            content = getattr(item, "content", None)
            if not isinstance(content, list):
                continue
            for part in content:
                text = getattr(part, "text", None)
                if isinstance(text, str) and text.strip():
                    chunks.append(text.strip())
        if chunks:
            return "\n".join(chunks).strip()

    raise ValueError("Judge returned no text output.")


def _parse_rubric_score(payload: str) -> RubricScore:
    try:
        parsed = json.loads(payload)
    except json.JSONDecodeError as exc:
        raise ValueError("Judge returned invalid JSON output.") from exc

    if not isinstance(parsed, dict):
        raise ValueError("Judge returned invalid JSON output.")

    return RubricScore.model_validate(parsed)


async def judge(
    rubric: Rubric,
    response: str,
    oai_client: openai.AsyncClient,
) -> RubricScore:
    """
    Judge a response against a rubric using OpenAI structured outputs.

    Args:
        rubric: Rubric definition used for judging.
        response: Response text or transcript to evaluate.
        oai_client: OpenAI async client used for structured output parsing.

    Returns:
        Parsed rubric judgment with per-dimension scores and pass/fail.
    """
    if not rubric.dimensions:
        raise ValueError("Cannot judge a rubric with no dimensions.")

    judge_config = _judge_config(rubric)
    schema = _judge_json_schema(rubric)
    try:
        api_response = await oai_client.responses.create(
            model=judge_config.model,
            instructions=_judge_instructions(rubric, schema),
            input=f"Response to evaluate:\n\n{response}",
            text={
                "format": {
                    "type": "json_schema",
                    "name": "rubric_score",
                    "description": "Structured rubric evaluation for an agent response.",
                    "schema": schema,
                    "strict": True,
                }
            },
            temperature=judge_config.temperature,
            max_output_tokens=judge_config.max_tokens,
        )
    except openai.AuthenticationError as exc:
        raise openai.AuthenticationError(
            message=(
                f"Judge authentication failed for model '{judge_config.model}'. "
                "Set a valid OPEN_ROUTER_API_KEY before running agentprobe."
            ),
            response=exc.response,
            body=exc.body,
        ) from exc

    score = _parse_rubric_score(_extract_response_text(api_response))
    score.validate_dimensions(rubric)
    return score


__all__ = [
    "JudgeDimensionScore",
    "RubricScore",
    "judge",
]

File: src/domains/evaluation/judge.ts
import { createHash } from "node:crypto";
import type { OpenAiResponsesClient } from "../../providers/sdk/openai-responses.ts";
import {
  OpenAiResponsesApiError,
  OpenAiResponsesAuthenticationError,
} from "../../providers/sdk/openai-responses.ts";
import type {
  JsonValue,
  OpenAiResponsesRequest,
  Rubric,
  RubricScore,
} from "../../shared/types/contracts.ts";
import { AgentProbeRuntimeError } from "../../shared/utils/errors.ts";

const DEFAULT_RETRY_DELAY_MS = 2000;
const MAX_JUDGE_ATTEMPTS = 3;
const JUDGE_SYSTEM_INSTRUCTIONS =
  "You are an expert rubric judge. Evaluate only the provided response using the supplied evaluation context.";

function judgeRetryDelayMs(): number {
  const raw = Bun.env.AGENTPROBE_JUDGE_RETRY_DELAY_MS?.trim();
  if (!raw) {
    return DEFAULT_RETRY_DELAY_MS;
  }
  const parsed = Number(raw);
  return Number.isFinite(parsed) && parsed >= 0
    ? parsed
    : DEFAULT_RETRY_DELAY_MS;
}

async function sleep(ms: number): Promise<void> {
  if (ms <= 0) {
    return;
  }
  await new Promise((resolve) => setTimeout(resolve, ms));
}

export function rubricToPromptMarkdown(rubric: Rubric): string {
  const lines = [
    `# Rubric: ${rubric.name}`,
    `- ID: \`${rubric.id}\``,
    `- Pass threshold: ${rubric.passThreshold.toFixed(2)}`,
  ];

  if (rubric.description) {
    lines.push(`- Description: ${rubric.description}`);
  }

  lines.push("", "## Dimensions");
  for (const dimension of rubric.dimensions) {
    lines.push(
      `### ${dimension.name}`,
      `- ID: \`${dimension.id}\``,
      `- Weight: ${dimension.weight.toFixed(2)}`,
      `- Scale type: ${dimension.scale.type}`,
    );
    if (dimension.scale.points !== undefined) {
      lines.push(`- Scale points: ${dimension.scale.points}`);
    }
    lines.push("- Scale labels:");
    for (const [score, label] of Object.entries(dimension.scale.labels)) {
      lines.push(`  - \`${score}\`: ${label}`);
    }
    lines.push("", "#### Judge Prompt", dimension.judgePrompt.trim(), "");
  }

  lines.push("## Scoring Overrides");
  const autoFail = rubric.scoringOverrides?.autoFailConditions ?? [];
  if (autoFail.length > 0) {
    lines.push("### Auto-Fail Conditions");
    for (const condition of autoFail) {
      lines.push(
        `- \`${condition.dimension}\`: ${
          condition.below !== undefined
            ? `below ${condition.below}`
            : `above ${condition.above}`
        }`,
      );
    }
  } else {
    lines.push("- Auto-fail conditions: none");
  }

  const autoPass = rubric.scoringOverrides?.autoPassConditions ?? [];
  if (autoPass.length > 0) {
    lines.push("", "### Auto-Pass Conditions");
    for (const condition of autoPass) {
      lines.push(
        `- \`${condition.dimension}\`: ${
          condition.below !== undefined
            ? `below ${condition.below}`
            : `above ${condition.above}`
        }`,
      );
    }
  } else {
    lines.push("- Auto-pass conditions: none");
  }

  lines.push("", "## Meta Prompt", rubric.metaPrompt.trim());
  return lines.join("\n");
}

function judgeJsonSchema(rubric: Rubric): Record<string, unknown> {
  const dimensionScoreSchema = {
    type: "object",
    properties: {
      reasoning: {
        type: "string",
        description: "Concise reasoning for the assigned score.",
      },
      evidence: {
        type: "array",
        items: { type: "string" },
        description:
          "Short evidence snippets or observations from the transcript.",
      },
      score: {
        type: "number",
        description: "Numeric score for the rubric dimension.",
      },
    },
    required: ["reasoning", "evidence", "score"],
    additionalProperties: false,
  };

  return {
    type: "object",
    properties: {
      dimensions: {
        type: "object",
        properties: Object.fromEntries(
          rubric.dimensions.map((dimension) => [
            dimension.id,
            {
              ...dimensionScoreSchema,
              description: `Score details for rubric dimension \`${dimension.id}\`.`,
            },
          ]),
        ),
        required: rubric.dimensions.map((dimension) => dimension.id),
        additionalProperties: false,
      },
      overall_notes: {
        type: "string",
        description: "Short overall summary of strengths and failures.",
      },
      pass: {
        type: "boolean",
        description: "Whether the evaluated response passes the rubric.",
      },
      failure_kind: {
        type: "string",
        enum: ["agent", "harness"],
        description:
          "When pass is false, classify whether the failure is due to the agent under test ('agent') or a defect in the test harness itself such as missing files or incorrect persona behavior ('harness'). When pass is true, set to 'agent'.",
      },
      failure_mode_detected: {
        type: ["string", "null"],
        description:
          "Named failure mode detected in the response, or null when none apply.",
      },
    },
    required: [
      "dimensions",
      "overall_notes",
      "pass",
      "failure_kind",
      "failure_mode_detected",
    ],
    additionalProperties: false,
  };
}

function judgeEvaluationContext(rubric: Rubric): string {
  const dimensionIds = rubric.dimensions
    .map((item) => item.id)
    .sort()
    .join(", ");
  return [
    rubricToPromptMarkdown(rubric),
    "",
    "Use the structured output schema attached to this request.",
    `The \`dimensions\` object must contain exactly these rubric dimension ids: ${dimensionIds}.`,
    "If the scenario defines failure modes, set `failure_mode_detected` to the matching named failure mode or null when none apply.",
  ].join("\n");
}

function judgePromptCacheKey(
  rubric: Rubric,
  evaluationContext: string,
  schema: Record<string, unknown>,
): string {
  const fingerprint = createHash("sha256")
    .update(rubric.id)
    .update("\n")
    .update(evaluationContext)
    .update("\n")
    .update(JSON.stringify(schema))
    .digest("hex")
    .slice(0, 16);
  return `agentprobe:judge:${rubric.id}:${fingerprint}`;
}

function judgeCacheControl(
  model: string,
): OpenAiResponsesRequest["cacheControl"] | undefined {
  return model.startsWith("anthropic/claude")
    ? { type: "ephemeral" }
    : undefined;
}

function parseRubricScore(payload: string): RubricScore {
  const parsed = JSON.parse(payload) as unknown;
  if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
    throw new AgentProbeRuntimeError("Judge returned invalid JSON output.");
  }

  const record = parsed as Record<string, unknown>;
  const dimensionsRaw = record.dimensions;
  if (
    !dimensionsRaw ||
    typeof dimensionsRaw !== "object" ||
    Array.isArray(dimensionsRaw)
  ) {
    throw new AgentProbeRuntimeError("Judge returned invalid JSON output.");
  }

  return {
    dimensions: Object.fromEntries(
      Object.entries(dimensionsRaw as Record<string, unknown>).map(
        ([key, value]) => {
          const item = value as Record<string, unknown>;
          return [
            key,
            {
              reasoning:
                typeof item.reasoning === "string" ? item.reasoning : "",
              evidence: Array.isArray(item.evidence)
                ? item.evidence.flatMap((entry) =>
                    typeof entry === "string" ? [entry] : [],
                  )
                : [],
              score:
                typeof item.score === "number"
                  ? item.score
                  : typeof item.score === "string"
                    ? Number(item.score)
                    : 0,
            },
          ];
        },
      ),
    ),
    overallNotes:
      typeof record.overall_notes === "string" ? record.overall_notes : "",
    passed: record.pass === true,
    failureKind:
      record.pass === false
        ? record.failure_kind === "harness"
          ? "harness"
          : "agent"
        : undefined,
    failureModeDetected:
      typeof record.failure_mode_detected === "string"
        ? record.failure_mode_detected
        : record.failure_mode_detected === null
          ? null
          : undefined,
  };
}

function validateRubricScore(rubric: Rubric, score: RubricScore): void {
  const expected = new Set(rubric.dimensions.map((dimension) => dimension.id));
  const actual = new Set(Object.keys(score.dimensions));
  const missing = [...expected].filter((id) => !actual.has(id));
  const extra = [...actual].filter((id) => !expected.has(id));
  if (missing.length > 0 || extra.length > 0) {
    const parts: string[] = [];
    if (missing.length > 0) {
      parts.push(`missing dimensions: ${missing.join(", ")}`);
    }
    if (extra.length > 0) {
      parts.push(`unexpected dimensions: ${extra.join(", ")}`);
    }
    throw new AgentProbeRuntimeError(
      `Judge output did not match rubric dimensions: ${parts.join("; ")}`,
    );
  }
}

export async function judgeResponse(
  rubric: Rubric,
  responseText: string,
  client: OpenAiResponsesClient,
): Promise<RubricScore> {
  if (rubric.dimensions.length === 0) {
    throw new AgentProbeRuntimeError(
      "Cannot judge a rubric with no dimensions.",
    );
  }
  if (!rubric.judge) {
    throw new AgentProbeRuntimeError("Rubric is missing judge configuration.");
  }
  if (rubric.judge.provider !== "openai") {
    throw new AgentProbeRuntimeError(
      `judge.ts only supports OpenAI judges, got: ${rubric.judge.provider}`,
    );
  }

  const schema = judgeJsonSchema(rubric);
  const evaluationContext = judgeEvaluationContext(rubric);
  const request: OpenAiResponsesRequest = {
    model: rubric.judge.model,
    instructions: JUDGE_SYSTEM_INSTRUCTIONS,
    input: [
      {
        type: "message",
        role: "user",
        content: [
          {
            type: "input_text",
            text: evaluationContext,
          },
        ],
      },
      {
        type: "message",
        role: "user",
        content: [
          {
            type: "input_text",
            text: `Response to evaluate:\n\n${responseText}`,
          },
        ],
      },
    ],
    text: {
      format: {
        type: "json_schema",
        name: "rubric_score",
        description: "Structured rubric evaluation for an agent response.",
        schema: schema as Record<string, JsonValue>,
        strict: true,
      },
    },
    temperature: rubric.judge.temperature,
    maxOutputTokens: rubric.judge.maxTokens,
    promptCacheKey: judgePromptCacheKey(rubric, evaluationContext, schema),
    cacheControl: judgeCacheControl(rubric.judge.model),
  };

  let lastError: unknown;
  for (let attempt = 1; attempt <= MAX_JUDGE_ATTEMPTS; attempt += 1) {
    try {
      const response = await client.create(request);
      if (!response.outputText.trim()) {
        throw new AgentProbeRuntimeError(
          "Judge response contained no text output.",
        );
      }
      const score = parseRubricScore(response.outputText);
      validateRubricScore(rubric, score);
      return score;
    } catch (error) {
      lastError = error;
      if (error instanceof OpenAiResponsesAuthenticationError) {
        throw new AgentProbeRuntimeError(
          `Judge authentication failed for model '${rubric.judge.model}'. Set a valid OPEN_ROUTER_API_KEY before running agentprobe.`,
        );
      }

      if (error instanceof OpenAiResponsesApiError) {
        const statusCode = error.statusCode;
        if (
          statusCode &&
          statusCode >= 400 &&
          statusCode < 500 &&
          statusCode !== 429
        ) {
          throw error;
        }
        if (attempt < MAX_JUDGE_ATTEMPTS) {
          await sleep(judgeRetryDelayMs());
          continue;
        }
        break;
      }

      if (
        error instanceof SyntaxError ||
        (error instanceof AgentProbeRuntimeError &&
          /invalid JSON output|no text output|did not match rubric dimensions/i.test(
            error.message,
          ))
      ) {
        if (attempt < MAX_JUDGE_ATTEMPTS) {
          continue;
        }
        break;
      }

      throw error;
    }
  }

  throw lastError instanceof Error
    ? lastError
    : new AgentProbeRuntimeError(String(lastError));
}
